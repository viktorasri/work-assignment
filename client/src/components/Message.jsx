import React from 'react';
import PropTypes from 'prop-types';

const Message = ({ children }) => {
  return <div className='message message--red'>{children}</div>;
};

Message.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
};

export default Message;
