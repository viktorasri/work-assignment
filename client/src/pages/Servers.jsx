import React from 'react';

import { ReactComponent as SortAsc } from '../assets/sort-asc.svg';
import { ReactComponent as SortDeasc } from '../assets/sort-deasc.svg';
import Page from '../components/Page';
import Button from '../components/Button';

const Servers = () => {
  return (
    <Page>
      <div className='servers'>
        <div className='row'>
          <div className='servers__heading'>
            <h1>Servers list</h1>
          </div>
          <table className='table'>
            <thead>
              <tr>
                <th>
                  <div className='d-flex align-center'>
                    <h3>Name</h3>
                    <span className='table__sort-btns'>
                      <Button variant='icon'>
                        <SortAsc className='icon icon--ice' />
                      </Button>
                      <Button variant='icon'>
                        <SortDeasc className='icon icon--ice' />
                      </Button>
                    </span>
                  </div>
                </th>
                <th>
                  <div className='d-flex align-center'>
                    <h3>Distance</h3>
                    <span className='table__sort-btns'>
                      <Button variant='icon'>
                        <SortAsc className='icon icon--ice' />
                      </Button>
                      <Button variant='icon'>
                        <SortDeasc className='icon icon--ice' />
                      </Button>
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>Server 1</td>
                <td>96</td>
              </tr>
              <tr>
                <td>Server 2</td>
                <td>545</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </Page>
  );
};

export default Servers;
