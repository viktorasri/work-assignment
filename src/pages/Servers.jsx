import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { ReactComponent as SortAsc } from '../assets/sort-asc.svg';
import { ReactComponent as SortDeasc } from '../assets/sort-deasc.svg';
import Page from '../components/Page';
import Button from '../components/Button';
import Spinner from '../components/Spinner';
import Message from '../components/Message';
import { getServersList, sortServersList } from '../redux/actions/serversActions';

const Servers = ({ history }) => {
  const dispatch = useDispatch();

  const userLogin = useSelector((state) => state.userLogin);
  const { error: errorLogin, success: successLogin } = userLogin;

  const serversList = useSelector((state) => state.serversList);
  const { error, loading, servers } = serversList;

  useEffect(() => {
    if (!successLogin) {
      history.push('/login');
    } else {
      dispatch(getServersList());
    }
  }, [successLogin, history, dispatch]);

  return (
    <Page>
      <div className='servers'>
        <div className='row'>
          <div className='servers__heading'>
            <h1>Servers list</h1>
          </div>
          {loading ? (
            <Spinner />
          ) : error ? (
            <Message>{error}</Message>
          ) : servers.length ? (
            <table className='table'>
              <thead>
                <tr>
                  <th>
                    <div className='d-flex align-center'>
                      <h3>Name</h3>
                      <span className='table__sort-btns'>
                        <Button variant='icon' onClick={() => dispatch(sortServersList('name', 'asc'))}>
                          <SortAsc className='icon icon--ice' />
                        </Button>
                        <Button variant='icon' onClick={() => dispatch(sortServersList('name', 'deasc'))}>
                          <SortDeasc className='icon icon--ice' />
                        </Button>
                      </span>
                    </div>
                  </th>
                  <th>
                    <div className='d-flex align-center'>
                      <h3>Distance</h3>
                      <span className='table__sort-btns'>
                        <Button variant='icon' onClick={() => dispatch(sortServersList('distance', 'asc'))}>
                          <SortAsc className='icon icon--ice' />
                        </Button>
                        <Button variant='icon' onClick={() => dispatch(sortServersList('distance', 'deasc'))}>
                          <SortDeasc className='icon icon--ice' />
                        </Button>
                      </span>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody>
                {servers.map((server, index) => (
                  <tr key={index}>
                    <td>{server.name}</td>
                    <td>{server.distance}</td>
                  </tr>
                ))}
              </tbody>
            </table>
          ) : (
            <Message variant='blue'>No servers available</Message>
          )}
        </div>
      </div>
    </Page>
  );
};

export default Servers;
