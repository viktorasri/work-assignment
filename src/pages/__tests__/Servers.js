import React from 'react';
import { shallow, mount } from 'enzyme';
import * as redux from 'react-redux';
import Servers from '../Servers';

import Spinner from '../../components/Spinner';
import Message from '../../components/Message';

describe('Page <Servers />', () => {
  const useSelectorSpy = jest.spyOn(redux, 'useSelector');
  const useDispatchSpy = jest.spyOn(redux, 'useDispatch');

  const mockServers = [
    { name: 'a', distance: 1 },
    { name: 'b', distance: 2 },
    { name: 'c', distance: 3 },
  ];

  it('expect to render component', () => {
    useSelectorSpy.mockReturnValue({ servers: mockServers, success: true });
    expect(shallow(<Servers />)).toMatchSnapshot();
  });

  it('expect to show spinner then loading', () => {
    useSelectorSpy.mockReturnValue({ loading: true });
    expect(shallow(<Servers />).find(Spinner).length).toEqual(1);
  });

  it('expect to show error message on login fail', () => {
    useSelectorSpy.mockReturnValue({ error: 'unauthorized' });
    expect(shallow(<Servers />).find(Message).length).toEqual(1);
  });

  it('should call history.push when user is not signed in', () => {
    const mockHistory = {
      push: jest.fn(),
    };
    const mockProps = {
      history: mockHistory,
    };
    useSelectorSpy.mockReturnValue({ success: false, servers: [] });
    const wrapper = mount(<Servers {...mockProps} />);
    expect(mockHistory.push).toHaveBeenCalledTimes(1);
  });
});
