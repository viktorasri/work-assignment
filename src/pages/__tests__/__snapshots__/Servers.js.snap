// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page <Servers /> expect to render component 1`] = `
<Page
  description="Server explorer web app"
>
  <div
    className="servers"
  >
    <div
      className="row"
    >
      <div
        className="servers__heading"
      >
        <h1>
          Servers list
        </h1>
      </div>
      <table
        className="table"
      >
        <thead>
          <tr>
            <th>
              <div
                className="d-flex align-center"
              >
                <h3>
                  Name
                </h3>
                <span
                  className="table__sort-btns"
                >
                  <Button
                    onClick={[Function]}
                    variant="icon"
                  >
                    <ForwardRef(SvgSortAsc)
                      className="icon icon--ice"
                    />
                  </Button>
                  <Button
                    onClick={[Function]}
                    variant="icon"
                  >
                    <ForwardRef(SvgSortDeasc)
                      className="icon icon--ice"
                    />
                  </Button>
                </span>
              </div>
            </th>
            <th>
              <div
                className="d-flex align-center"
              >
                <h3>
                  Distance
                </h3>
                <span
                  className="table__sort-btns"
                >
                  <Button
                    onClick={[Function]}
                    variant="icon"
                  >
                    <ForwardRef(SvgSortAsc)
                      className="icon icon--ice"
                    />
                  </Button>
                  <Button
                    onClick={[Function]}
                    variant="icon"
                  >
                    <ForwardRef(SvgSortDeasc)
                      className="icon icon--ice"
                    />
                  </Button>
                </span>
              </div>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            key="0"
          >
            <td>
              a
            </td>
            <td>
              1
            </td>
          </tr>
          <tr
            key="1"
          >
            <td>
              b
            </td>
            <td>
              2
            </td>
          </tr>
          <tr
            key="2"
          >
            <td>
              c
            </td>
            <td>
              3
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</Page>
`;
